File: problems/weakvertices/weakvertices.py
while True:
    n = int(input())

    if n == -1:
        break

    matrix = []

    for i in range(n):
        row = list(map(int, input().split()))
        matrix.append(row)
    
    failed = set()
    for start in range(n):
        opts = set()
        found = False

        for first in range(n):
            if matrix[start][first]:
                opts.add(first)
        
        opts2 = set()
        for first in opts:
            for second in range(n):
                if matrix[first][second]:
                    opts2.add(second)
        
        for second in opts2:
            if matrix[second][start]:
                found = True
                break
        
        if not found:
            failed.add(start)
    
    for f in sorted(list(failed)):
        print(f, end=" ")
    print()
        

File: problems/weakvertices/weakvertices.cpp
#include 
#include 
#include 
#include 

using namespace std;

int main() {
    int n;
    cin >> n;
    
    while (n != -1) {
        
        vector> graph;
        
        for (int i = 0; i < n; i++) {
            vector tmp;
            for (int j = 0; j < n; j++) {
                int x;
                cin >> x;
                tmp.push_back(x);
            }
            graph.push_back(tmp);
        }
        
        unordered_set failed;
        
        for (int start = 0; start < n; start++) {
            
            bool found = false;
            
            unordered_set options1;
            for (int second = 0; second < n; second++) {
                if (graph[start][second]) {
                    options1.emplace(second);
                }
            }
            
            unordered_set options2;
            for (auto second : options1) {
                for (int third = 0; third < n; third++) {
                    if (graph[second][third]) {
                        options2.emplace(third);
                    }
                }
            }
            
            for (auto third : options2) {
                if (graph[third][start]) {
                    found = true;
                    break;
                }
            }
            if (!found) {
                failed.emplace(start);
            }
            
            
        }
        
        vector output;
            for (auto i : failed) {
                output.push_back(i);
            }
            sort(output.begin(), output.end());
            
            for (int i = 0; i < output.size(); i++) {
                cout << output[i] << " ";
            }
            cout << endl;
        
        cin >> n;
    }
    
    return 0;
}